src/main/scala/tilelink/Nodes.scala

```diff
@@ -32,17 +32,52 @@ object TLImp extends NodeImp[TLClientPortParameters, TLManagerPortParameters, TL
     pu.copy(managers = pu.managers.map { m => m.copy (nodePath = node +: m.nodePath) })
 }
 
-case class TLClientNode(portParams: Seq[TLClientPortParameters])(implicit valName: ValName) extends SourceNode(TLImp)(portParams)
-case class TLManagerNode(portParams: Seq[TLManagerPortParameters])(implicit valName: ValName) extends SinkNode(TLImp)(portParams)
+trait TLFormatNode extends BaseNode
+{
+  def edges: Edges[List[TLEdgeIn], List[TLEdgeOut]]
+  override def formatNode() = "TileLink Node\n" + edges.out.map(currEdge =>
+  currEdge.client.clients.map(currClient => "Output Edges (edges.out.map.client.clients.map):" +
+  "\nClient Name = " + currClient.name +
+  "\nvisibility = " + currClient.visibility +
+  (if (currClient.knownToEmit == None) "\nEmits parameters are UNKNOWN\n" else
```

Comment:
I think you want some tabs in the scala code here and maybe also in the output string. You might consider looking at the `"""` escape for multiline strings with `stripMargin` (https://alvinalexander.com/scala/how-to-create-multiline-strings-heredoc-in-scala-cookbook) and/or using string interpolation.